src/methods.jl

```diff
@@ -10,318 +19,378 @@ derivatives.
-parameters.
-
-The recognized keywords are:
-
-* `adapt`: The number of adaptive steps to use improve the estimate of `bound`.
-* `bound`: Bound on the value of the function and its derivatives at `x`.
-* `condition`: The condition number. See [`DEFAULT_CONDITION`](@ref).
-* `eps`: The assumed roundoff error. Defaults to `add_tiny(eps())`.
-
-!!! warning
-    Bounds can't be adaptively computed over nonstandard grids; passing a value for
-    `adapt` greater than 0 when `m::Nonstandard` results in an error.
-
-!!! note
-    Calling [`FiniteDifferenceMethod`](@ref) objects is equivalent to passing them to `fdm`.
-
-# Examples
-
-```julia-repl
-julia> fdm(central_fdm(5, 1), sin, 1; adapt=2)
-0.5403023058681039
-
-julia> fdm(central_fdm(2, 1), exp, 0, Val(true))
-(FiniteDifferenceMethod:
-  order of method:       2
-  order of derivative:   1
-  grid:                  [-1, 1]
-  coefficients:          [-0.5, 0.5]
-  roundoff error:        1.42e-14
-  bounds on derivatives: 1.00e+02
-  step size:             1.69e-08
-  accuracy:              1.69e-06
-, 1.0000000031817473)
-```
+    function estimate_step(
+        m::FiniteDifferenceMethod,
+        f::Function,
+        x::T;
+        factor::Int=1,
+        max_step=convert(T, 0.1)
+    ) where T<:AbstractFloat
+
+Estimate the step size for a finite difference method `m`. Also estimates the error of the
+estimate of the derivative.
+
+# Arguments
+- `m::FiniteDifferenceMethod`: Finite difference method to estimate the step size for.
+- `f::Function`: Function to evaluate the derivative of.
+- `x::T`: Point to estimate the derivative at.
+
+# Keywords
+- `factor::Int=1`. Factor to amplify the estimated round-off error by. This can be used
+    to force a more conservative step size.
+- `max_step=convert(T, 0.1)`: Maximum step size.
+
+# Returns
+- `Tuple{T, T}`: Estimated step size and an estimate of the error of the finite difference
+    estimate.
 """
-function fdm(
-    m::M,
-    f,
-    x::T,
-    ::Val{true};
-    condition=DEFAULT_CONDITION,
-    bound=_estimate_bound(f(x), condition),
-    eps=add_tiny(Base.eps(bound)),
-    adapt=m.history.adapt,
-    max_step=convert(T, 0.1),
-) where {T<:AbstractFloat, M<:FiniteDifferenceMethod}
-    if M <: Nonstandard && adapt > 0
-        throw(ArgumentError("can't adaptively compute bounds over Nonstandard grids"))
-    end
-    eps > 0 || throw(ArgumentError("eps must be positive, got $eps"))
-    bound > 0 || throw(ArgumentError("bound must be positive, got $bound"))
-    0 <= adapt < 20 - m.p || throw(ArgumentError("can't perform $adapt adaptation steps"))
-
-    # The below calculation can fail for `m.q == 0`, because then `C₂` may be zero. We
-    # therefore handle this edge case here.
-    m.q == 0 && return m, f(x)
-
-    p = m.p
+function estimate_step(
+    m::FiniteDifferenceMethod,
+    f::Function,
+    x::T;
+    factor::Int=1,
+    max_step=convert(T, 0.1)
```

Comment:
This is a pretty strong assumption about the units of `x`…  If you're going to make this assumption, then I would just pass `1` or `0.1` to `Richardson.extrapolate`.

Reply:
Perhaps there's a better way to go about this. Limiting the step size is mainly for functions with a non-differentiability, like `abs`. I agree that `0.1` is pretty arbitrary and depends on the scale of `x`, and maybe we can choose it differently, e.g. capping it by a multiple of `abs(f(x))`.

My main worry with passing a fixed value like `1` or `0.1` is that is may be too large. Example:

```julia
julia> extrapolate(1, rtol=0) do h
           @show h
           (inv(1e-3 + h) - inv(1e-3)) / h
       end
h = 1.0
h = 0.125
h = 0.015625
(-8927.580356151784, 7928.579357150786)

julia> extrapolate(0.1, rtol=0) do h
           @show h
           (inv(1e-3 + h) - inv(1e-3)) / h
       end
h = 0.1
h = 0.0125
h = 0.0015625
(-83241.65749908325, 73340.66740007335)

julia> extrapolate(0.01, rtol=0) do h
           @show h
           (inv(1e-3 + h) - inv(1e-3)) / h
       end
h = 0.01
h = 0.00125
h = 0.00015625
h = 1.953125e-5
h = 2.44140625e-6
h = 3.0517578125e-7
h = 3.814697265625e-8
h = 4.76837158203125e-9
h = 5.960464477539063e-10
(-1.000000000029157e6, 3.0059716664254665e-5)
```

With the current implementation,

```julia
julia> extrapolate_fdm(forward_fdm(2, 1), inv, 1e-3)
(-1.0000000000231842e6, 3.273598849773407e-6)
```

An option would be to set the default cap to `Inf` and have the user set it whenever it is required.

Reply:
> e.g. capping it by a multiple of abs(f(x)).

This is just as as bad if not worse — it assumes that `f(x)` and `x` have the same units.

Reply:
Every number you choose will be wrong for some function, of course…

(The basic problem with `inv(x)` at `x=1e-3` is that the radius of convergence of the Taylor series is so small at that point, of course.  Hence `extrapolate` is stopping because it detects a lack of convergence.   It's possible that I could use a better stopping criterion that somehow distinguishes between a lack of convergence because of floating-point cancellation — where we really should stop — and a lack of convergence due to the Taylor series — where should continue decreasing the step.)

Reply:
> Every number you choose will be wrong for some function, of course…

Of course.

An important use case of the package is to automatically check the implementation of a derivative by comparing it with a numerical estimate. Since there are many rules to check, you really want to not bother appropriately configuring a finite difference for every one of them. For such use cases, it's important that the initialisation is as robust as it can be.

Regarding `max_step`, perhaps the two best options are to default it to `Inf` or to something like `max(maximum(abs, x), 0.1 * oneunit(x))`, which scales with `x`. Would you be happy with any of these?

Reply:
`oftype(x, 0.1 * max(abs(x), oneunit(real(x))))` seems as good as anything, I suppose.